doc/ni_by_example_4.md

```diff
@@ -0,0 +1,169 @@
+
+* `w`: Append column to stream
+  * `$ ni <data> w[np'a*a']`
+  * `w` will add columns only up to the length of the input stream
+* `W`: Prepend column stream
+  * `$ ni <data> Wn` - Add line numbers to the stream (by prepending one element the infinite stream `n`)
+  * `W` will add rows only up to the length of the input stream
+* `v`: Vertical operation on columns
+  * **Important Note**: As of 2016-12-23, this operator is too slow to use in production.
+  
+##Advanced Row and Column Operations
+* `j` - streaming join
+  * Note that this join will consume a single line of both streams; it does **NOT** provide a SQL-style left or right join.
+* `Y` - dense-to-sparse transformation
+  * Explodes each row of the stream into several rows, each with three columns:
+    * The index of the row that the input data that came from
+    * The index of the column that the input data came from
+    * The value of the input stream at the row + column specified by the first two columns.
+* `X` - sparse-to-dense transformation
+  * `X` inverts `Y`; it converts a specifically-formatted 3-column stream into a multiple-column stream.
+  * The specification for what the input matrix must look like is described above in the `Y` operator.
+  
+  
+
+##Things other than Perl 
+
+Look, these are here, and if it helps you get started with `ni`, great. But `ni` is written in Perl, for Perl, and in a Perlic style. Use these, but go learn Perl.
+
+*  `m'<...>'`: Ruby
+   * applies the Ruby snippet `<...>` to each row of the stream 
+*  `l'<...>'`: Lisp
+   * applies the Lisp snippet `<...>` to each row of the stream 
+
+##Understanding the `ni` monitor
+More details [here](monitor.md). Overall:
+
+* Negative numbers = non-rate-determining step
+* Positive numbers = rate-determining step
+* Large Positive numbers = slow step
+
+##Cell Operations 
+`$ ni <data> ,<op><columns>`
+
+These provide keystroke-efficient ways to do transformations on a single column of the input data. Of particular use is the deterministic hashing function, which does a good job of compacting long IDs into 32-bit integers. With ~40 million IDs, there will be only be about 1% hash collisions, and with 400 million IDs, there will be 10% hash collisions.  See [this](http://math.stackexchange.com/questions/35791/birthday-problem-expected-number-of-collisions) for why.
+
+* `,a`: Running average
+* `,d`: Difference between consecutive rows
+* `,e`: Natural exponential (`e**x`)
+* `,h`: Murmurhash (deterministic 32-bit hash function)
+* `,j<amt>`: Jitter (add uniform random noise in the range `[-amt/2, amt/2]`)
+* `,l`: Natural log (`ln x`)
+* `,s`: Running sum 
+* `,q`: Quantize
+* `,z`: Intify (hash and then convert hash values to integers starting with 1)
+
+
+
+##Connecting `bash` and `ni`
+
+* `e[<script>]`: Evaluate script
+  * evaluate `<script>` in bash, and stream out the results one line at a time.
```

Comment:
Technically, "pipe the current stream through the command" -- you could do `ni n10 e[tac]` for example